src/core/highlight.ts

```diff
@@ -24,32 +40,31 @@ import type { Grammar, GrammarToken, GrammarTokens, RegExpLike } from '../types'
 export function highlight (
 	this: Prism,
 	text: string,
-	language: string,
+	languageRef: string | LanguageLike,
 	options: HighlightOptions = {}
 ): string {
 	const prism = this ?? singleton;
 
-	const languageId = prism.components.resolveAlias(language);
-	const grammar = options.grammar ?? prism.languageRegistry.getLanguage(languageId);
+	const { id, def, language } = prism.languageRegistry.resolveRef(languageRef);
 
-	const env: Record<string, any> = {
+	let env: HookEnv['after-tokenize'] = {
 		code: text,
-		grammar,
+		languageId: id,
+		languageDef: def,
 		language,
 	};
 
 	prism.hooks.run('before-tokenize', env);
 
-	if (!env.grammar) {
-		throw new Error('The language "' + env.language + '" has no grammar.');
+	if (!env.language) {
+		throw new Error(`No language definition found for ${env.languageId}.`);
 	}
 
-	env.tokens = prism.tokenize(env.code, env.grammar.resolvedGrammar);
+	env = env as unknown as HookEnv['after-tokenize'];
```

Comment:
It's only for TS, right? As an alternative, I'd suggest explicitly adding a `token` property to `env`:

```suggestion
	env = { ...env, tokens: undefined } as HookEnv['after-tokenize'];
```